File: skills/code-architect/references/code-examples/go/decorator.go
// Decorator: attach extra responsibilities to an object dynamically by
// wrapping it in objects sharing the same interface.
package main

import "fmt"

// Component is the interface shared by the core and all decorators.
type Component interface {
	Process(input string) string
}

// core is the base behaviour being decorated.
type core struct{}

func (core) Process(input string) string { return input }

// trimDecorator wraps a component and adds trimming-like framing.
type trimDecorator struct {
	inner Component
}

func (d trimDecorator) Process(input string) string {
	return "<" + d.inner.Process(input) + ">"
}

// repeatDecorator wraps a component and duplicates the result.
type repeatDecorator struct {
	inner Component
}

func (d repeatDecorator) Process(input string) string {
	out := d.inner.Process(input)
	return out + out
}

func main() {
	var c Component = core{}
	c = trimDecorator{inner: c}
	c = repeatDecorator{inner: c}
	fmt.Println(c.Process("data"))
}

File: skills/code-architect/references/code-examples/csharp/Decorator.cs
// Decorator — attach responsibilities by wrapping objects at runtime.
using System;

namespace Patterns.Decorator
{
    interface IComponent
    {
        string Operation();
    }

    sealed class ConcreteComponent : IComponent
    {
        public string Operation() => "core";
    }

    abstract class Decorator : IComponent
    {
        protected readonly IComponent Inner;
        protected Decorator(IComponent inner) => Inner = inner;
        public abstract string Operation();
    }

    sealed class UpperDecorator : Decorator
    {
        public UpperDecorator(IComponent inner) : base(inner) { }
        public override string Operation() => Inner.Operation().ToUpperInvariant();
    }

    sealed class BracketDecorator : Decorator
    {
        public BracketDecorator(IComponent inner) : base(inner) { }
        public override string Operation() => $"[{Inner.Operation()}]";
    }

    static class Demo
    {
        public static void Run()
        {
            IComponent plain = new ConcreteComponent();
            Console.WriteLine(plain.Operation());

            // Stack decorators dynamically; order matters.
            IComponent wrapped = new BracketDecorator(new UpperDecorator(plain));
            Console.WriteLine(wrapped.Operation());
        }

        public static void Main() => Run();
    }
}
